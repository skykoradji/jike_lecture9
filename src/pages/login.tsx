import React, { Component, Fragment } from 'react';
import { Button, Grid, Typography } from '@material-ui/core';
import { AxiosResponse } from 'axios';
import { TextValidator, ValidatorForm } from 'react-material-ui-form-validator';
import { withStyles, createStyles } from '@material-ui/core/styles';
import { Theme } from '@material-ui/core';
import withLayout from '../lib/withLayout';
import restClient from '../lib/restClient';
import errorHandler from '../lib/errorHandler';

const styles = (theme: Theme) => createStyles({
  loginContainer: {
    width: 500,
    margin: '0 auto',
    paddingLeft: 15,
    paddingRight: 15,
    textAlign: 'center',
    [theme.breakpoints.down('sm')]: {
      width: 500
    },
    [theme.breakpoints.down('xs')]: {
      width: '100%'
    }
  },
  loginButton: {
    float: 'right',
    marginTop: 40
  },
  loginForm: {
    marginTop: 40
  },
  clear: {
    clear: 'both'
  },
  headline: {
    marginTop: 30,
    textAlign: 'center'
  },
  signupSection: {
    textAlign: 'center',
    marginTop: 10
  },
  signupInLogin: {
    marginLeft: 10
  },
  hr: {
    border: 0,
    borderTop: '1px solid #f5f5f5'
  }
});

interface Props {
  name: string;
  history: any;
  classes: any;
  updateUser: (payload: any) => void;
  authenticate: (payload: any) => void;
}

interface State {
  username: string;
  password: string;
  submitted: boolean;
}

class LoginPage extends Component<Props, State> {
  constructor(props: any) {
    super(props);
    this.state = {
      username: '',
      password: '',
      submitted: false
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  async handleSubmit() {
    const { username, password } = this.state;

    this.setState({ submitted: true });
    try {
      const response: AxiosResponse<any> = await restClient().post('signin', { userName: username, password });

      const { token, role, uniqueId } = response.data;
      
      const { authenticate, updateUser, history } = this.props;
      localStorage.setItem('token', token);
      localStorage.setItem('uniqueId', uniqueId);
      authenticate(response.data);
      const profile: AxiosResponse<any> = await restClient().get(`userprofile/${uniqueId}`);

      updateUser({ ...profile.data, role });

      history.push('/');
    } catch (err) {
      errorHandler(err);
      this.setState({ submitted: false });
    }
  }

  signup = () => {
    const { history } = this.props;
    history.push('/signup');
  };

  handleNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ username: event.target.value });
  }

  handlePasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ password: event.target.value });
  }

  render() {
    const { classes } = this.props;
    const { submitted, username, password } = this.state;
    return (
      <Fragment>
        <div className={classes.loginContainer}>
          <Typography className={classes.headline} variant="h4" component="h4">
            Login
          </Typography>

          <Grid container justify="center">
            <Grid item xs={12}>
              <ValidatorForm className={classes.loginForm} onSubmit={this.handleSubmit}>
                <TextValidator
                  required
                  fullWidth
                  margin="normal"
                  label="Username"
                  onChange={this.handleNameChange}
                  name="userName"
                  disabled={submitted}
                  value={username}
                  validators={['required', 'minStringLength:5']}
                  errorMessages={[
                    'This field is required!',
                    'Username cannot be shorter than 5 characters!'
                  ]}
                />

                <TextValidator
                  required
                  fullWidth
                  margin="normal"
                  label="Password"
                  onChange={this.handlePasswordChange}
                  name="password"
                  type="password"
                  validators={['required', 'minStringLength:5']}
                  errorMessages={[
                    'This field is required',
                    'Password cannot be shorter than 8 characters!'
                  ]}
                  value={password}
                />

                <Button
                  className={classes.loginButton}
                  variant="contained"
                  color="primary"
                  type="submit"
                  disabled={submitted}
                >
                  Login
                </Button>
                <div className={classes.clear} />
                <hr className={classes.hr} />
              </ValidatorForm>
            </Grid>
          </Grid>
        </div>
        <div className={classes.signupSection}>
          Donâ€™t have an account?
          <Button
            className={classes.signupInLogin}
            variant="text"
            color="primary"
            size="small"
            onClick={this.signup}
          >
            Sign up
          </Button>
        </div>
      </Fragment>
    );
  }
}

export const Login = withStyles(styles)(LoginPage);
export default withLayout(Login);
