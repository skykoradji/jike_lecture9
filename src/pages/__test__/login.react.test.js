/**
 * @jest-environment jsdom
 */
import React from 'react';
import Enzyme, { shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Login } from '../login';

Enzyme.configure({ adapter: new Adapter() });
describe('Login Component', () => {
  it('renders login form', () => {
    expect(shallow(<Login />).contains('loginForm'));
  });
});

test('Login component snapshot', () => {
  const shallowed = shallow(<Login />);
  expect(shallowed).toMatchSnapshot();
});

test('Login component snapshot', () => {
  const tree = render(<Login />);
  expect(tree).toMatchSnapshot();
});
