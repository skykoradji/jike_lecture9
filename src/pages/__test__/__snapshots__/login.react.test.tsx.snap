// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component snapshot 1`] = `
<LoginPage
  classes={
    Object {
      "clear": "LoginPage-clear-4",
      "headline": "LoginPage-headline-5",
      "hr": "LoginPage-hr-8",
      "loginButton": "LoginPage-loginButton-2",
      "loginContainer": "LoginPage-loginContainer-1",
      "loginForm": "LoginPage-loginForm-3",
      "signupInLogin": "LoginPage-signupInLogin-7",
      "signupSection": "LoginPage-signupSection-6",
    }
  }
/>
`;

exports[`Login component snapshot 2`] = `
Array [
  <div
    class="LoginPage-loginContainer-1"
  >
    <h4
      class="MuiTypography-root-9 MuiTypography-h4-24 LoginPage-headline-5"
    >
      Login
    </h4>
    <div
      class="MuiGrid-container-45 MuiGrid-justify-xs-center-62"
    >
      <div
        class="MuiGrid-item-46 MuiGrid-grid-xs-12-85"
      >
        <form
          class="LoginPage-loginForm-3"
        >
          <div
            class="MuiFormControl-root-142 MuiFormControl-marginNormal-143 MuiFormControl-fullWidth-145"
          >
            <label
              class="MuiFormLabel-root-157 MuiFormLabel-required-162 MuiInputLabel-required-150 MuiInputLabel-root-146 MuiInputLabel-formControl-151 MuiInputLabel-animated-154"
              data-shrink="false"
            >
              Username
              <span
                class="MuiFormLabel-asterisk-163"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root-177 MuiInput-root-164 MuiInput-underline-168 MuiInputBase-fullWidth-186 MuiInput-fullWidth-171 MuiInputBase-formControl-178 MuiInput-formControl-165"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input-187 MuiInput-input-172"
                name="userName"
                required=""
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root-142 MuiFormControl-marginNormal-143 MuiFormControl-fullWidth-145"
          >
            <label
              class="MuiFormLabel-root-157 MuiFormLabel-required-162 MuiInputLabel-required-150 MuiInputLabel-root-146 MuiInputLabel-formControl-151 MuiInputLabel-animated-154"
              data-shrink="false"
            >
              Password
              <span
                class="MuiFormLabel-asterisk-163"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root-177 MuiInput-root-164 MuiInput-underline-168 MuiInputBase-fullWidth-186 MuiInput-fullWidth-171 MuiInputBase-formControl-178 MuiInput-formControl-165"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input-187 MuiInput-input-172 MuiInputBase-inputType-190 MuiInput-inputType-175"
                name="password"
                required=""
                type="password"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root-220 MuiButton-root-194 MuiButton-contained-205 MuiButton-containedPrimary-206 MuiButton-raised-208 MuiButton-raisedPrimary-209 LoginPage-loginButton-2"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label-195"
            >
              Login
            </span>
          </button>
          <div
            class="LoginPage-clear-4"
          />
          <hr
            class="LoginPage-hr-8"
          />
        </form>
      </div>
    </div>
  </div>,
  <div
    class="LoginPage-signupSection-6"
  >
    Don’t have an account?
    <button
      class="MuiButtonBase-root-220 MuiButton-root-194 MuiButton-text-196 MuiButton-textPrimary-197 MuiButton-flat-199 MuiButton-flatPrimary-200 MuiButton-sizeSmall-217 LoginPage-signupInLogin-7"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-195"
      >
        Sign up
      </span>
    </button>
  </div>,
]
`;
